iceoryx_examples/icedelivery/iox_publisher_typed_experimental.cpp

```diff
@@ -0,0 +1,87 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "topic_data.hpp"
+
+#include "iceoryx_posh/experimental/popo/publisher.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+
+#include <chrono>
+#include <iostream>
+#include <thread>
+
+bool killswitch = false;
+
+static void sigHandler(int f_sig [[gnu::unused]])
+{
+    // caught SIGINT, now exit gracefully
+    killswitch = true;
+}
+
+void getVehiclePosition(Position* allocation, uint64_t multiplier)
+{
+    new (allocation) Position(1111.1111 * multiplier, 1111.1111 * multiplier, 1111.1111 * multiplier);
+}
+
+int main(int argc, char* argv[])
+{
+    // Register sigHandler for SIGINT
+    signal(SIGINT, sigHandler);
+
+    iox::runtime::PoshRuntime::getInstance("/iox-ex-publisher-modern");
+
+    auto typedPublisher = iox::popo::TypedPublisher<Position>({"Odometry", "Position", "Vehicle"});
+    typedPublisher.offer();
+
+    float_t ct = 0.0;
+    while (!killswitch)
+    {
+        ++ct;
+
+        // Retrieve a typed sample from shared memory.
+        // Sample can be held until ready to publish.
+        auto result = typedPublisher.loan();
+        if (!result.has_error())
+        {
+            auto& sample = result.get_value();
+            sample->x = ct * 1.1;
+            sample->y = ct * 1.1;
+            sample->z = ct * 1.1;
+            typedPublisher.publish(sample);
+        }
+
+        // Retrieve a sample and provide logic to immediately populate and publish via a lambda.
+        typedPublisher.loan().and_then([&](iox::popo::PublishableSample<Position>& sample) {
+            auto allocation = sample.get();
+            // Do some stuff leading to eventually generating the data in the provided sample's shared memory...
+            new (allocation) Position(ct * 11.11, ct * 11.11, ct * 11.11);
+            // ...then publish the bytes
+            sample.publish();
+        });
+
+        // Simple copy-and-publish. Useful for smaller data types.
+        auto position = Position(ct * 111.111, ct * 111.111, ct * 111.111);
+        typedPublisher.publishCopyOf(position);
+
+        // Samples can be generated within any callable and written directly to the loaned memory
+        // allocation. The first argument of the callable must be T*, this will point to the memory
+        // allocation.
+        typedPublisher.publishResultOf(getVehiclePosition, ct);
+        typedPublisher.publishResultOf([](Position* allocation) { new (allocation) Position(0, 0, 0); });
```

Comment:
Very good. Accepts any general callable entity it seems. Avoids copies by writing directly to the provided memory. I would try to avoid mixing Position and allocation like this. Either we are concrete and deal with positions at the callable level, or we are agnostic to the actual data and think in memory terms such as allocation.

Reply:
Position is just the particular type being published here, that could be Foo* allocation.